File: leetcode/prefix_sum/go/product_of_array_except_self.go
package _go

func productExceptSelf(nums []int) []int {
	result := make([]int, len(nums))
	prefix := 1
	for i := 0; i < len(nums); i++ {
		result[i] = prefix
		prefix *= nums[i]
	}
	postfix := 1
	for i := len(nums) - 1; i >= 0; i-- {
		result[i] *= postfix
		postfix *= nums[i]
	}
	return result
}

// / This seems a standard way of prefix sum approach
func productExceptSelf2(nums []int) []int {
	result := make([]int, len(nums))
	result[0] = 1
	result[len(nums)-1] = 1
	postfix := 1
	for i := 1; i < len(nums); i++ {
		result[i] = result[i-1] * nums[i-1]
	}
	for i := len(nums) - 2; i >= 0; i-- {
		postfix *= nums[i+1]
		result[i] *= postfix
	}
	return result
}

File: leetcode/prefix_sum/product_of_array_except_self.dart
/**
 * Product of an array except itself demonstration
 *
 * This example using multiply arithmetic
 *
 * nums = [1,2,3,4]
 * output = [24,12,8,6]
 *
 * // Explanation
 *  nums[0] = 1 => output[0] = nums[1] * nums[2] * nums[3] = 2 * 3 * 4 = 24
 *  nums[1] = 2 => output[1] = nums[0] * nums[2] * nums[3] = 1 * 3 * 4 = 12
 *  nums[2] = 3 => output[2] = nums[0] * nums[1] * nums[3] = 1 * 2 * 4 = 8
 *  nums[3] = 4 => output[3] = nums[0] * nums[1] * nums[2] = 1 * 2 * 3 = 6
 *
 * // Draft
 * First calculate the result of the nums by multiply all the element in the
 * `nums` array, then use that result to divide by index specific index
 * output[0] = 1 * 2 * 3 * 4 = 24 / nums[0] = 24 / 1 = 24
 * output[1] = 1 * 2 * 3 * 4 = 24 / nums[1] = 24 / 2 = 12
 * output[2] = 1 * 2 * 3 * 4 = 24 / nums[2] = 24 / 3 = 8
 * output[3] = 1 * 2 * 3 * 4 = 24 / nums[3] = 24 / 4 = 6
 *
 */

class Solution {
  List<int> productExceptSelf(List<int> nums) {
    List<int> res = List<int>.filled(nums.length, 1);
    int prefix = 1;
    for (int i = 0; i < nums.length; i++) {
      res[i] = prefix;
      prefix *= nums[i];
    }
    int postfix = 1;
    for (int i = nums.length - 1; i >= 0; i--) {
      res[i] *= postfix;
      postfix *= nums[i];
    }
    return res;
  }
}

void main(List<String> args) {
  final nums = [1, 2, 3, 4];
  final obj = Solution();
  print(obj.productExceptSelf(nums));
}
